BL-5056 Handle captive portal better

gmartin7 opened the pull request:
On my developer machine, triggering the captive portal code results in a Yellow screen error that the user can "Continue" from. If I understand the NonFatalProblem.Report() params correctly, on a Beta or higher release it would just "toast".
This only handles the version upgrade case, unfortunately there are several other areas where Bloom accesses the internet that are not handled here.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1869)
<!-- Reviewable:end -->

StephenMcConnel commented:
Generally looks okay, but I don't understand how the BloomUpdateWebClient class can possibly work without an explicit implementation of the interface method IBloomWebClient.DownloadString(string)

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/BloomUpdateWebClient.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvSshtA12HHVp0C3Gpy:-KvSshtBLChOK75dfnr5:baladt2) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/BloomUpdateWebClient.cs#L15)):*
> ```Smalltalk
> 	/// <summary>
> 	/// Allows moq-ing of DownloadString to return a simulated captive portal.
> 	/// </summary>
> ```

How is DownloadString implemented when it's not moq-ed?  Is it somehow automatically exposed from WebClient? The code compiles okay but I don't understand how since there's no implementation given for the interface method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-:-KvSz-LdtoBEJqbmRW5J:b-nm1bwz)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/BloomExe/BloomUpdateWebClient.cs, line 15 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvSshtA12HHVp0C3Gpy:-KvT1uIefiLedlJpoT1B:b2irzy0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/BloomUpdateWebClient.cs#L15)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

How is DownloadString implemented when it's not moq-ed?  Is it somehow automatically exposed from WebClient? The code compiles okay but I don't understand how since there's no implementation given for the interface method.
</blockquote></details>

Yes, DownloadString(string) is a WebClient method.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Okay, the interface method exposes an existing class method in a way that can overridden for testing.  I guess that makes sense.  <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for what this PR does.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-:-KvT3wBU0MYHUTU6sfjM:b-mpww7l)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Reviewed 3 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/BloomUpdateWebClient.cs, line 9 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYUoRTR9ehXonLQlMG:-KvYUoRTR9ehXonLQlMH:bmqrp3w) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/BloomUpdateWebClient.cs#L9)):*
> BloomUpdateWebClient

Would this class be used for other interactions other than update if someone can get that working? If so, how about BloomWebClient?

---

*[src/BloomExe/UpdateVersionTable.cs, line 56 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYVXOQxbzZlQbS80zA:-KvYVXOQxbzZlQbS80zB:bxrb8t6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L56)):*
> must have been an error

Returning a non-null result is how we indicate error? Or is this comment misplaced/incorrect?

If non-null is the indication of error, how can we refactor that to make it less confusing?
An idea (but not sure I like it): have the method return a bool indicating success with an out parameter containing the error details.

---

*[src/BloomExe/UpdateVersionTable.cs, line 106 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYVknFn_wFZGT4ST3P:-KvYVknFn_wFZGT4ST3Q:bloo6ta) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L106)):*
> ```Smalltalk
> 		}
> 
> 		/// <summary>
> ```

The comment should indicate what the method actually does.
Indicating it is factored this way for testing should be supplemental.

---

*[src/BloomExe/UpdateVersionTable.cs, line 125 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYY6_wAYlny2kMwjaq:-KvYY6_wAYlny2kMwjar:b9d3ljr) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L125)):*
> Beta or higher release it would just "toast"

> NonFatalProblem.Report(ModalIf.Beta, PassiveIf.All, msg); // hopefully this will just 'toast'

This code would make Beta (and Alpha/dev) present a yellow screen (ModalIf.Beta).

---

*[src/BloomExe/UpdateVersionTable.cs, line 152 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYWMCtyl5g-lfZB5lj:-KvYWMCtyl5g-lfZB5lk:bkyvimc) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L152)):*
> //html may have javascript which has braces which will kill the string.format in WriteEvent

This comment goes with the line with curly braces. It confused me here.

---

*[src/BloomTests/UpdateVersionTableTests.cs, line 166 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYXLuX8ZbNw2BIKnJv:-KvYXLuX8ZbNw2BIKnJw:b9jhrpk) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomTests/UpdateVersionTableTests.cs#L166)):*
> Assert.That(t.LookupURLOfUpdateInternal(client), Is.Null);

As I indicated in the comments above, LookupURLOfUpdateInternal returning null when it "Works" is completely counterintuitive to me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Reviewed 2 of 5 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
Review status: 1 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/UpdateVersionTable.cs, line 56 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYVXOQxbzZlQbS80zA:-Kvc6RZfUWHvQfxPX24w:b-40luqb) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L56)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> must have been an error

Returning a non-null result is how we indicate error? Or is this comment misplaced/incorrect?

If non-null is the indication of error, how can we refactor that to make it less confusing?
An idea (but not sure I like it): have the method return a bool indicating success with an out parameter containing the error details.
</blockquote></details>

Yeah, I think I like that better. Much clearer.

---

*[src/BloomExe/UpdateVersionTable.cs, line 106 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYVknFn_wFZGT4ST3P:-Kvc75htZpmazERc68Pr:b-v2mofl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L106)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

The comment should indicate what the method actually does.
Indicating it is factored this way for testing should be supplemental.
</blockquote></details>

Done. Renamed 'internal' method to CanGetVersionTableFromWeb().

---

*[src/BloomExe/UpdateVersionTable.cs, line 125 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYY6_wAYlny2kMwjaq:-Kvc8DLBH8dqEgRaRuNL:b-y0sn75) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L125)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> Beta or higher release it would just "toast"

> NonFatalProblem.Report(ModalIf.Beta, PassiveIf.All, msg); // hopefully this will just 'toast'

This code would make Beta (and Alpha/dev) present a yellow screen (ModalIf.Beta).
</blockquote></details>

Thanks for the clarification. Changed to ModalIf.Alpha.

---

*[src/BloomExe/UpdateVersionTable.cs, line 152 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYWMCtyl5g-lfZB5lj:-Kvc8gPrRYMxMMptvp6n:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/UpdateVersionTable.cs#L152)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> //html may have javascript which has braces which will kill the string.format in WriteEvent

This comment goes with the line with curly braces. It confused me here.
</blockquote></details>

Done.

---

*[src/BloomTests/UpdateVersionTableTests.cs, line 166 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYXLuX8ZbNw2BIKnJv:-Kvc8m6z_Kp35dRdy3uu:b8brqv0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomTests/UpdateVersionTableTests.cs#L166)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> Assert.That(t.LookupURLOfUpdateInternal(client), Is.Null);

As I indicated in the comments above, LookupURLOfUpdateInternal returning null when it "Works" is completely counterintuitive to me.
</blockquote></details>

Fixed. And the changes here also resulted in the tests being better, I think.

---

*[src/BloomExe/BloomUpdateWebClient.cs, line 9 at r1](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-KvYUoRTR9ehXonLQlMG:-Kvc4xPNwcX4_CjnpFIf:b-kocnp6) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/854179f1b7cf24cc9a28ed17c5d4754e0c4ec822/src/BloomExe/BloomUpdateWebClient.cs#L9)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

> BloomUpdateWebClient

Would this class be used for other interactions other than update if someone can get that working? If so, how about BloomWebClient?
</blockquote></details>

Quite possibly. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1869#-:-KvcPufwa5nHpZUSAPta:bnfp4nl)*
<!-- Sent from Reviewable.io -->
